File: Intermediate-Algorithm-Scripting-Js-Py/4-Wherefor-Art-Thou.py
def whatIsInAName(collection, source):
    return [thing for thing in collection
            if all(key_val in thing.items() for key_val in source.items())]

# === tests ===
print(whatIsInAName([{ 'first': "Romeo", 'last': "Montague" }, { 'first': "Mercutio", 'last': None }, { 'first': "Tybalt", 'last': "Capulet" }], { 'last': "Capulet" }))

File: Intermediate-Algorithm-Scripting-Js-Py/4-Wherefor-Art-Thou.js
function whatIsInAName(collection, source) {
    return collection.filter(obj => {
        return Object.keys(source).every(key => source[key] === obj[key])
    })
}


// === tests ===
let result = whatIsInAName(
    [
        { first: "Romeo", last: "Montague" }, 
        { first: "Mercutio", last: null }, 
        { first: "Tybalt", last: "Capulet" }
    ], 
    { last: "Capulet" })

console.log(result) 
